File: packages/go/core/headers.go
package core

import (
	"fmt"
	"strings"
	"time"
)

type DeprecationHeaders map[string]string

func dateToUnix(isoDate string) int64 {
	t, _ := time.Parse("2006-01-02", isoDate)
	return t.UTC().Unix()
}

func dateToHTTPDate(isoDate string) string {
	t, _ := time.Parse("2006-01-02", isoDate)
	return t.UTC().Format("Mon, 02 Jan 2006 15:04:05") + " GMT"
}

func BuildHeaders(ep EndpointConfig) DeprecationHeaders {
	h := DeprecationHeaders{
		"Deprecation": fmt.Sprintf("@%d", dateToUnix(ep.DeprecatedAt)),
	}
	if ep.SunsetAt != nil {
		h["Sunset"] = dateToHTTPDate(*ep.SunsetAt)
	}
	var links []string
	if ep.Successor != nil {
		links = append(links, fmt.Sprintf(`<%s>; rel="successor-version"`, *ep.Successor))
	}
	if ep.MigrationDoc != nil {
		links = append(links, fmt.Sprintf(`<%s>; rel="deprecation"`, *ep.MigrationDoc))
	}
	if len(links) > 0 {
		h["Link"] = strings.Join(links, ", ")
	}
	return h
}

func DaysUntilSunset(ep EndpointConfig) *int {
	if ep.SunsetAt == nil {
		return nil
	}
	sunset, err := time.Parse("2006-01-02", *ep.SunsetAt)
	if err != nil {
		return nil
	}
	days := int(time.Until(sunset).Hours() / 24)
	return &days
}

File: packages/javascript/src/core/headers.ts
import { EndpointConfig } from './models'

export interface DeprecationHeaders {
  Deprecation: string
  Sunset?: string
  Link?: string
}

function toUnix(isoDate: string): number {
  return Math.floor(new Date(isoDate + 'T00:00:00Z').getTime() / 1000)
}

function toHttpDate(isoDate: string): string {
  return new Date(isoDate + 'T00:00:00Z').toUTCString()
}

export function buildHeaders(endpoint: EndpointConfig): DeprecationHeaders {
  const headers: DeprecationHeaders = {
    Deprecation: `@${toUnix(endpoint.deprecated_at)}`,
  }

  if (endpoint.sunset_at) {
    headers.Sunset = toHttpDate(endpoint.sunset_at)
  }

  const links: string[] = []
  if (endpoint.successor) {
    links.push(`<${endpoint.successor}>; rel="successor-version"`)
  }
  if (endpoint.migration_doc) {
    links.push(`<${endpoint.migration_doc}>; rel="deprecation"`)
  }
  if (links.length) {
    headers.Link = links.join(', ')
  }

  return headers
}

export function daysUntilSunset(endpoint: EndpointConfig): number | null {
  if (!endpoint.sunset_at) return null
  const sunset = new Date(endpoint.sunset_at).getTime()
  return Math.floor((sunset - Date.now()) / 86_400_000)
}
